pkg/usagetracker/tracker.go

```diff
@@ -0,0 +1,929 @@
+			Name: "cortex_usage_tracker_snapshot_cleanups_failed_files_total",
+			Help: "Total number of files that failed to be deleted during snapshots cleanup.",
+		})
+		t.snapshotCleanupsDeletedFiles = promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_usage_tracker_snapshot_cleanups_deleted_files_total",
+			Help: "Total number of snapshots deleted during snapshots cleanup.",
+		})
+		t.snapshotsRemainingInTheBucket = promauto.With(registerer).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_usage_tracker_snapshots_remaining_in_the_bucket",
+			Help: "Number of snapshots remaining in the bucket.",
+		})
+		t.snapshotsRemainingInTheBucket.Set(-1) // We don't know yet.
+	}
+
+	t.Service = services.NewBasicService(t.start, t.run, t.stop)
+	return t, nil
+}
+
+func (t *UsageTracker) shouldPerformCleanups() bool {
+	return t.instanceID == 0
+}
+
+// start implements services.StartingFn.
+func (t *UsageTracker) start(ctx context.Context) error {
+	// Start dependencies.
+	t.subservicesWatcher = services.NewFailureWatcher()
+	t.subservicesWatcher.WatchService(t.instanceLifecycler)
+	if err := services.StartAndAwaitRunning(ctx, t.instanceLifecycler); err != nil {
+		return errors.Wrap(err, "unable to start instance lifecycler")
+	}
+
+	return nil
+}
+
+func (t *UsageTracker) run(ctx context.Context) error {
+	// TODO: check here if all partitions already have owners, in which case we should reconcilePartitions immediately (no need to wait).
+	// If there are no owners yet, this is a cold start, so wait until all instances have joined the ring to avoid re-shuffling.
+
+	var snapshotsCleanupTickerChan <-chan time.Time
+	if t.shouldPerformCleanups() {
+		ticker := time.NewTicker(util.DurationWithJitter(t.cfg.SnapshotCleanupInterval, t.cfg.SnapshotCleanupIntervalJitter))
+		defer ticker.Stop()
+		snapshotsCleanupTickerChan = ticker.C
+	}
+
+	for {
+		select {
+		case <-time.After(t.cfg.PartitionReconcileInterval):
+			if err := t.reconcilePartitions(ctx); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-t.subservicesWatcher.Chan():
+			return errors.Wrap(err, "usage-tracker dependency failed")
+		case <-snapshotsCleanupTickerChan:
+			if err := t.cleanupSnapshots(ctx); err != nil {
+				level.Error(t.logger).Log("msg", "failed to clean up old snapshots", "err", err)
+			}
+		}
+	}
+}
+
+func (t *UsageTracker) reconcilePartitions(ctx context.Context) error {
+	// We're going to hold the read lock for the whole function, as we're the only function that can write (we assume stop() can't be called because we're still running)
+	// We'll take the write mutex when we want to modify t.partitions.
+	t.partitionsMtx.RLock()
+	defer t.partitionsMtx.RUnlock()
+	// locked will run f while write-locked
+	locked := func(f func()) {
+		t.partitionsMtx.RUnlock()
+		t.partitionsMtx.Lock()
+		f()
+		t.partitionsMtx.Unlock()
+		t.partitionsMtx.RLock()
```

Comment:
Is it OK if another goroutine read-locks in between line 376 and 377? If yes, then why take the read-lock at all? As the comment says, no other goroutine is going to write to `t.partitions`, so we can safely read from it in this function/goroutine.